### DISCUSSION TITLE
Call c# method from button on xyzDialog.ts

### DISCUSSION THREAD

**Discussion by elenaaralla** [serenity-is/Serenity#6840] [General]
Hi all,

I'm looking for a simple sample on how to call a c# function to perform a connection test; at the moment I added the button, that display an info dialog.

I want that the button calls MyClass.MyMethod(xyzRow.Id) and use the value returned from this call to display a message:

```
protected getToolbarButtons() {
        let buttons = super.getToolbarButtons();

        buttons.push({
            title: localText(Texts.Site.PecMailboxDialog.TestConnection),
            cssClass: 'edit-permissions-button',
            icon: 'fa-check text-green',
            onClick: () =>
            {

                // var res = MyClass.MyMethod(xyzRow.Id);

                informationDialog(
                    res,
                    () => {
                        return;
                    });
            }
        });

        return buttons;
    }
```

How can I do?

**Comment by wezmag** [comment]
1. Create `TestConnectionRequest` class derives from `ServiceRequest`.

```csharp
public class TestConnectionRequest : ServiceRequest
{
    public int Id { get; set; }
}
```

2. Create `TestConnectionResponse` class derives from `ServiceResponse`.

```csharp
public class TestConnectionResonse : ServiceResponse
{
    public string Message { get; set; }
}
```

3. Add TestConnection method in your XyzEndpoint.cs file.

```csharp
[HttpPost]
[ServiceAuthorize(PermissionKeys.Security)] //for permission control, change the permission keys accordingly.
public TestConnectionResponse TestConnection(TestConnectionRequest request)
{
    //do connection test here

    return new TestConnectionResponse
    {
        Message = "Successful!"
    };
}
```

4. Call backend service on button click.

```typescript
onClick: () => {
    const request: TestConnectionRequest = { Id: this.entity.Id };
    XyzService.TestConnection(request, response => {
        if (response.Message)
            informationDialog(response.Message, () => {});
    });
}
```

**Reply by elenaaralla** [reply]
Hi @wezmag,

I've done as you suggested:

All works fine; I can compile and run the code without problem; the only strange thing is that Visual Studio signals me an error (see attached image)

Thank you so much! 

![Schermata 2023-07-19 alle 10 31 34](https://github.com/serenity-is/Serenity/assets/14870650/27a0c27f-1367-472e-b8b8-42da39d7cb8d)

**Reply by wezmag** [reply]
On top of the typescript file, add `import { TestConnectionRequest } from "@/ServerTypes/YourModule"`.
Replace `@/ServerTypes/YourModule` accordingly.
